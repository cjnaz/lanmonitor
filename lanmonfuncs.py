#!/usr/bin/env python3
"""LAN monitor support functions
"""

# __version__ = "V1.0a 210515"

#==========================================================
#
#  Chris Nelson, 2021
#
# V1.1  210523  cmd timeout tweaks
# V1.0  210507  V1.0
# V1.0a 210515  Set timeouts to 1s for ping and 5s for ssh commands on remotes
#
# Changes pending
#   
#==========================================================

import sys
import os.path
import subprocess
import datetime
import time
import re
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../funcs3/'))
from funcs3 import logging, getcfg, snd_email, snd_notif

# Configs / Constants
NOTIF_SUBJ = "LAN Monitor"
RTN_PASS     = 0
RTN_WARNING  = 1
RTN_FAIL     = 2
RTN_CRITICAL = 3


# Module global vars
args = None
keylen = 0      # Padding vars used for pretty printing
hostlen = 0


def convert_time (timeX):
    """ Given time value term such as 90s, 15m, 20h, 3d, 2w return int seconds and units.
        <1h> returns (3600, "hours")
    If no units suffix, the value is interpreted as secs.
        <90> returns (90, "secs ")
    """
    if type(timeX) is int:
        return timeX, "secs "           # Case Int
    try:
        return int(timeX), "secs "      # Case Str without units
    except:
        pass
    try:                                # Case Str with units
        time_value = int(timeX[:-1])
        time_unit =  timeX[-1:].lower()
        if time_unit == "s":
            return time_value, "secs "
        if time_unit == "m":
            return time_value * 60, "mins "
        if time_unit == "h":
            return time_value * 60*60, "hours"
        if time_unit == "d":
            return time_value * 60*60*24, "days "
        if time_unit == "w":
            return time_value * 60*60*24*7, "weeks"
        raise ValueError(f"Illegal time units <{time_unit}>")
    except Exception as e:
        logging.error (f"ERROR:  Could not convert time value <{timeX}>\n  {e}")
        raise


USER_HOST_FORMAT = re.compile(r"[\w.-]+@([\w.-]+)$")
USER_HOST_PORT_FORMAT = re.compile(r"[\w.-]+@([\w.-]+):([\d]+)$")

def split_user_host_port(u_h_p):
    """ Handle variations in passed-in user_host_port (u_h_p)
        "local", "user@host", or "user@host:port"
    Return separate user_host, host, and port values
    EG:
        "xyz@host15:4455" returns:
            ["xyz@host15", "host15", "4455"]
        "xyz@host15" returns:
            ["xyz@host15", "host15", "22"]  (the default port is "22" for ssh)
        "local" returns
            ["local", "local", ""]
    """
    user_host_noport = u_h_p.split(":")[0]
    host = u_h_p
    port = ""
    if host != "local":
        port = "22"
        out = USER_HOST_FORMAT.match(u_h_p)
        if out:
            host = out.group(1)
        else:
            out = USER_HOST_PORT_FORMAT.match(u_h_p)
            if out:
                host = out.group(1)
                port = out.group(2)
            else:
                _msg = f"Expecting <user@host> or <user@host:port> format, but found <{u_h_p}>."
                logging.error(f"ERROR:  {_msg}")
                raise ValueError (_msg)
    return user_host_noport, host, port


def cmd_check(cmd, user_host_port, return_type=None, check_line_text=None, expected_text=None, not_text=None):
    """  Runs the cmd and operates on the response based on return_type.
    return_types:
        check_string        Returns True if expected_text occurs in response of the cmd, plus the full full subprocess run structure
            check_line_text     If provided, only the first line containing this text is checked
            expected_text       Text that must be found
            not_text            Text that must NOT be found
        cmdrun              Returns True if the cmd return code was 0, plus the full full subprocess run structure
    """

    if user_host_port != "local":
        u_h, _, port = split_user_host_port(user_host_port)
        cmd = ["ssh", u_h, "-p" + port, "-o", "ConnectTimeout=1", "-T"] + cmd

    for nTry in range (getcfg('nRetries')):
        try:
            logging.debug(f"cmd_check subprocess command try {nTry+1}: <{cmd}>")
            # runtry = subprocess.run(cmd, capture_output=True, text=True)  # Py 3.7+
            runtry = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)   #Py3.6 requires old-style params
        except Exception as e:
            logging.error(f"ERROR:  subprocess.run of cmd <{cmd}> failed.\n  {e}")
            return False, None

        if return_type == "check_string":
            if check_line_text is None:
                text_to_check = runtry.stdout
            else:
                text_to_check = ""
                for line in runtry.stdout.split("\n"):
                    if check_line_text in line:
                        text_to_check = line
                        break
                
            if expected_text in text_to_check:
                if not_text is not None:
                    if not_text not in text_to_check:
                        return True, runtry
                else:
                    return True, runtry
            else:
                if nTry == getcfg('nRetries')-1:
                    return False, runtry

        elif return_type == "cmdrun":
            if runtry.returncode == 0:
                return True, runtry
            elif nTry == getcfg('nRetries')-1:
                return False, runtry

        else:
            _msg = f"Invalid return_type <{return_type}> passed to cmd_check"
            logging.error (f"ERROR:  {_msg}")
            raise ValueError (_msg)
        time.sleep (convert_time(getcfg('RetryInterval'))[0])


have_access_dict = {}

def have_access(user_host_port):
    """If user_host_port != "local" then check ping and ssh access to the host.
    Remember passing result and avoid redoing the ping and ssh tests on later checks.
    If ping/ssh tests fail they will be retried if referenced in a later monitored item.
    The dictionary is cleared on each RecheckInterval in the lanmonitor main code.
    ping or ssh access fails return a tuple of (False, error_message_string).
    (True, "") is returned if local or if ping and ssh tests pass for non-local host.

    NOTE that this optimization can result in confusion:
        if access once passed then fails the plugin will be called and may flag an error
        while the actual problem is no access.
    """
    user_host_port = user_host_port.lower()
    if user_host_port == "local":
        return (True, "")
    if user_host_port in have_access_dict:
        return (True, "")

    _, host, _ = split_user_host_port(user_host_port)
    if host != "local":     # Check for remote access if non-local
        pingrslt = cmd_check(["ping", "-c", "1", "-W", "1", host], user_host_port="local", return_type="cmdrun")
        if not pingrslt[0]:
            return (False, "HOST CANNOT BE REACHED")
        else:
            sshrslt = cmd_check(["echo", "hello"], user_host_port=user_host_port, return_type="cmdrun")
            if "Connection refused" in sshrslt[1].stderr:
                return (False, f"SSH CONNECTION REFUSED <{user_host_port}>")
            # if "The authenticity of host" in sshrslt[1].stdout:       # Never happens thru the script, only interactive
            #     return (False, f"SSH CONNECTION NOT APPROVED.  ssh to <{user_host_port}> manually first.")

    have_access_dict[user_host_port] = True
    return (True, "")


def next_summary_timestring():
    """Calculate timestring of next summary.
    SummaryDays			0 1 2 3 4 5 6	# Days of week.  0 = Sunday
    SummaryTime			9:45		    # 24 hour clock
    NOTE:  May be off by 1 hour over a DTS changes.
    """
    try:
        target_hour   = int(getcfg("SummaryTime","").split(":")[0])
        target_minute = int(getcfg("SummaryTime","").split(":")[1])
        today_day_num = datetime.datetime.today().isoweekday()    # Sunday = 0, Saturday = 6
        now = datetime.datetime.now().replace(second=0, microsecond=0)
        next_summary = now + datetime.timedelta(days=30)
        for daynum in getcfg("SummaryDays").split():
            offset_num_days = int(daynum) - today_day_num
            if offset_num_days < 0:
                offset_num_days += 7
            plus_day =  now + datetime.timedelta(days=offset_num_days)
            with_time = plus_day.replace(hour=target_hour, minute=target_minute)
            if with_time < datetime.datetime.now():
                with_time = with_time + datetime.timedelta(days=7)
            if with_time < next_summary:
                next_summary = with_time
        logging.debug(f"Next summary:  {next_summary}")
        return next_summary
    except Exception as e:
        _msg = f"SummaryDays <{getcfg('SummaryDays','')}> or SummaryTime <{getcfg('SummaryTime','')}> settings could not be parsed\n  {e}"
        logging.error(f"ERROR:  {_msg}")
        raise ValueError (_msg) from None
        