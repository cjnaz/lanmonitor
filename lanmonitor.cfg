# LAN Monitor config file
#  220420  Updated to V1.4 (LogLevel and LogFile)


# Core tool params
nRetries            3               # Num times to try to read an item before erroring
RetryInterval       1s              # Wait time between failed item reads
StartupDelay        30s             # Wait time when starting in service mode to allow everything to come up fully at system boot before checking items
RecheckInterval     15m             # Wait time between main loops
Gateway             192.168.1.1     # IP or Hostname used for checking for network access as a gate for checks on non-local hosts
LogLevel            30              # Logging module levels: 10:DEBUG, 20:INFO, 30:WARNING (default), 40:ERROR, 50:CRITICAL
LogFile             <path/to>/logfile.txt    # absolute or relative (to script directory) path


# Whitespace separated list of notification handlers
Notif_handlers      stock_notif

# Stock notifications handler params
CriticalReNotificationInterval 	2h      # Wait time for critical event notification message resend
SummaryDays             0 1 2 3 4 5 6   # Days of week.  0 = Sunday, 6 = Saturday
SummaryTime             09:00           # 24 hour clock
LogSummary              True            # True: Force logging of summary


# Monitor types and items
MonType_SELinux		selinux_plugin
SELinux_local           local           enforcing       # or permissive

MonType_Host		pinghost_plugin
Host_Printer_Server     local           192.168.1.44

MonType_Service	    service_plugin
Service_lanmonitor      local           lanmonitor 
Service_routermonitor   local           routermonitor 
Service_wanstatus       local           wanstatus
Service_httpd           local           httpd 
Service_firewalld       local           CRITICAL    firewalld 
Service_smb	            local           CRITICAL    smb 
Service_sshd            local           sshd
Service_mariadb         local           mariadb
Service_RPi1_cups       me@RPi1.lan     CRITICAL    cups

MonType_Page        webpage_plugin
Page_WeeWX              local           http://localhost/weewx/         Current Conditions
Page_xBrowserSync       me@RPi2.mylan   https://www.xbrowsersync.org/   Browser syncing as it should be: secure, anonymous and free! # Check it out!

MonType_Process     process_plugin
Process_x11vnc          local           CRITICAL    /usr/bin/x11vnc

MonType_Activity    fsactivity_plugin
Activity_MyServer_backups local         8d     /mnt/share/MyServerBackups
Activity_RPi2_log.csv   rpi2.mylan      5m     /mnt/RAMDRIVE/log.csv

MonType_YumUpdate  yum_update_history_plugin
YumUpdate_myserver      local           CRITICAL  15d  update --skip-broken


# Email and Notification params
EmailFrom	        your.email@gmail.com
NotifList           4805551234@vzwpix.com   # For notifications - One or more addresses, whitespace separated.  Use your carrier's email-to-text bridge address.  Regular email addresses may be used.
EmailTo             your.email@gmail.com    # For perodic summaries - One or more addresses, whitespace separated.
Import              ~/creds_SMTP            # Provides EmailServer, EmailServerPort, EmailUser and EmailPass
#EmailVerbose	    True                    # True: enable the emailer debug level
#DontEmail          True                    # True: Emails (including notifications) will NOT be sent
#DontNotif          True                    # True: Notifications will NOT be sent
