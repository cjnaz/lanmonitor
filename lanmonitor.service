[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path to>/lanmonitor

[Install]
WantedBy=multi-user.target
