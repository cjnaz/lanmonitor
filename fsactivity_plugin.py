#!/usr/bin/env python3
"""LAN Monitor plugin - fsactivity_plugin
"""

__version__ = "V0.0 210415"

#==========================================================
#
#  Chris Nelson, 2021
#
# V0.0 210415  Initial
#
# Changes pending
#   
#==========================================================

import sys
import datetime
import re
import lanmonfuncs
from funcs3 import logging  #, cfg, getcfg

# Configs / Constants
RTN_PASS     = 0
RTN_WARNING  = 1
RTN_FAIL     = 2
RTN_CRITICAL = 3
LSMATCH = re.compile(r'[ldrwx\-.]+\s+[\d*]\s+[\w\d]+\s+[\w\d]+\s+[\d]+\s+([\d\-]+)+\s([\d:.]+)\s([\d\-]+)')
    # Given -rwxrw-r--. 1 cjn users     976269 2021-03-11 00:30:44.049364380 -0700 filexyz.txt
    # group(1) == 2021-03-11, group(2) == 21:24:27.269508428, group(3) == -0700


class monitor:

    def __init__ (self):
        pass

    def eval_status (self, item):
        """ Primary function for checking the status of this item type.
        Passed in item dictionary keys:
            key             Full 'itemtype_tag' key value from config file line
            keylen          string length of longest key of this item type
            tag             'tag' portion only from 'itemtype_tag' from config file line
            user_host       'local' or 'user@hostname' from config file line
            host            'local' or 'hostname' from config file line
            hostlen         string length of longest host of this item type
            critical        True if 'CRITICAL' is in the config file line
            rest_of_line    Remainder of line after the 'user_host' from the config file line

        Returns dictionary with these keys:
            rslt            Integer status:  RTN_PASS, RTN_WARNING, RTN_FAIL, RTN_CRITICAL
            notif_key       Unique handle for tracking active notifications in the notification handler 
            message         String with status and context details
        """
        
        # Construct item type specifics and check validity
        key = item["key"]
        key_padded = key.ljust(item["keylen"])
        host = item["host"]
        host_padded = host.ljust(item["hostlen"])
        xx = item["rest_of_line"].split(maxsplit=1)
        maxage = xx[0]
        maxage_sec, units = lanmonfuncs.convert_time(maxage)     # aborts on conversion error
        path = xx[1]

        # Check for remote access if non-local
        if host != "local":
            pingrslt = lanmonfuncs.cmd_check(["ping", host, "-c", "1"], user_host="local", return_type="cmdrun")
            if not pingrslt[0]:
                return {"rslt":RTN_WARNING, "notif_key":key, "message":f"WARNING: {key} - {host} - HOST CANNOT BE REACHED"}

        # Process the item
        cmd = ["ls", "-ltA", "--full-time", path]
        ls_rslt = lanmonfuncs.cmd_check(cmd, user_host=item['user_host'], return_type="cmdrun")
        # print (rslt)

        if not ls_rslt[0]:
            return {"rslt":RTN_WARNING, "notif_key":key, "message":f"WARNING: {key} - {host} - COULD NOT GET ls OF PATH <{path}>"}
        ls_list = ls_rslt[1].stdout.split("\n")
        newest_file = ls_list[0]                # When path is to a file
        if newest_file.startswith("total"):
            newest_file = ls_list[1]            # When path is to a directory
        out = LSMATCH.match(newest_file)
        if out:
            xx = f"{out.group(1)} {out.group(2)[0:8]} {out.group(3)}"   # 2021-03-11 00:30:44 -0700
            newest_timestamp = (datetime.datetime.strptime(xx, "%Y-%m-%d %H:%M:%S %z")).timestamp()
            newest_age = datetime.datetime.now().timestamp() - newest_timestamp
        else:
            return {"rslt":RTN_WARNING, "notif_key":key, "message":f"WARNING: {key} - {host} - COULD NOT GET TIMESTAMP OF NEWEST FILE <{path}>"}

        def retime(time_sec):
            """ Convert time value back to original units """
            if units == "secs ":  return time_sec
            if units == "mins ":  return time_sec /60
            if units == "hours":  return time_sec /60/60
            if units == "days ":  return time_sec /60/60/24
            if units == "weeks":  return time_sec /60/6024/7

        if newest_age < maxage_sec:
            return {"rslt":RTN_PASS, "notif_key":key, "message":f"{key_padded}  OK - {host_padded} - {retime(newest_age):6.1f} {units} ({int(retime(maxage_sec)):>4} {units} max)  {path}"}
        else:
            if item["critical"]:
                return {"rslt":RTN_CRITICAL, "notif_key":key, "message":f"CRITICAL: {key}  STALE FILES - {host} - {retime(newest_age):6.1f} {units} ({int(retime(maxage_sec)):>4} {units} max)  {path}"}
            else:
                return {"rslt":RTN_FAIL, "notif_key":key, "message":f"FAIL: {key}  STALE FILES - {host} - {retime(newest_age):6.1f} {units} ({int(retime(maxage_sec)):>4} {units} max)  {path}"}


if __name__ == '__main__':
    import argparse
    from funcs3 import loadconfig

    CONFIG_FILE = "lanmonitor.cfg"

    parser = argparse.ArgumentParser(description=__doc__ + __version__, formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('--config-file', default=CONFIG_FILE,
                        help=f"Path to config file (default <{CONFIG_FILE}>).")
    parser.add_argument('-V', '--version', action='version', version='%(prog)s ' + __version__,
                        help="Return version number and exit.")

    lanmonfuncs.args = parser.parse_args()
    loadconfig(cfgfile=lanmonfuncs.args.config_file)

    inst = monitor()
 
    test = {"key":"Activity_Shop2_backups", "keylen":20, "tag":"Shop2_backups","host":"local", "user_host":"local", "hostlen":10, "critical":False, "rest_of_line":"8d /mnt/share/backups/Shop2/"}
    print(f"{test}\n  {inst.eval_status(test)}\n")
    test = {"key":"Activity_rpi2_ramdrive", "keylen":20, "tag":"rpi2_ramdrive","host":"rpi2.lan", "user_host":"pi@rpi2.lan", "hostlen":10, "critical":False, "rest_of_line":"5m		/mnt/RAMDRIVE"}
    print(f"{test}\n  {inst.eval_status(test)}\n")
    test = {"key":"Activity_Shop2_backups_fail", "keylen":20, "tag":"Shop2_backups_fail","host":"local", "user_host":"local", "hostlen":10, "critical":False, "rest_of_line":"1h /mnt/share/backups/Shop2/"}
    print(f"{test}\n  {inst.eval_status(test)}\n")
    test = {"key":"Activity_badpath", "keylen":20, "tag":"badpath","host":"local", "user_host":"local", "hostlen":10, "critical":False, "rest_of_line":"1h /mnt/share/xxx"}
    print(f"{test}\n  {inst.eval_status(test)}\n")
    test = {"key":"Activity_badhost", "keylen":20, "tag":"badhost","host":"rpi3.lan", "user_host":"pi@rpi3.lan", "hostlen":10, "critical":False, "rest_of_line":"1h /mnt/share/xxx"}
    print(f"{test}\n  {inst.eval_status(test)}\n")

    sys.exit()