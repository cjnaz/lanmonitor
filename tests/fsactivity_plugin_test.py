#!/usr/bin/env python3
"""LAN Monitor plugin - fsactivity_plugin_test
"""

#==========================================================
#
#  Chris Nelson, Copyright 2021-2023
#
# 3.0 230301 - Packaged
#
#==========================================================

from cjnfuncs.cjnfuncs import set_toolname, setuplogging, logging, cfg
from lanmonitor.lanmonfuncs import RTN_PASS
try:
    from fsactivity_plugin import monitor
except:
    from lanmonitor.fsactivity_plugin import monitor

tool = set_toolname("tool")
cfg["nRetries"]         = 1
cfg["RetryInterval"]    = "0s"
cfg["ConsoleLogFormat"] = "{module:>35}.{funcName:20} - {levelname:>8}:  {message}"
setuplogging()
logging.getLogger().setLevel(logging.DEBUG)

test_num = 0
def dotest (test):
    global test_num
    test_num += 1
    print(f"\nTest {test_num} {'-'*50}")
    inst = monitor()
    setup_rslt = inst.setup(test)
    logging.debug (f"{test['key']} - setup() returned:  {setup_rslt}")
    if setup_rslt == RTN_PASS:
        logging.debug (f"{test['key']} - eval_status() returned:  {inst.eval_status()}")

dotest ({"key":"Activity_Shop2_backups", "tag":"Shop2_backups", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"8d /mnt/share/backups/Shop2/"})

dotest ({"key":"Activity_rpi3_ramdrive", "tag":"rpi3_ramdrive", "host":"rpi3", "user_host_port":"pi@rpi3", "critical":True, "check_interval":1, "rest_of_line":"5m		/mnt/RAMDRIVE"})

dotest ({"key":"Activity_Shop2_backups_fail", "tag":"Shop2_backups_fail", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"1h /mnt/share/backups/Shop2/"})

dotest ({"key":"Activity_empty_dir", "tag":"empty_dir", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"1h junk"}) # empty dir in cwd

dotest ({"key":"Activity_badpath", "tag":"badpath", "host":"local", "user_host_port":"local", "critical":True, "check_interval":1, "rest_of_line":"1h /mnt/share/xxx"})

dotest ({"key":"Activity_badhost", "tag":"badhost", "host":"rpi3.lan", "user_host_port":"pi@rpi3.lan", "critical":True, "check_interval":1, "rest_of_line":"1h /mnt/share/xxx"})
