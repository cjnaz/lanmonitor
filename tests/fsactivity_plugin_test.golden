
Test 1 - Local activity pass - OK --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_local_dir_pass - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_local_dir_pass', 'tag': 'local_dir_pass', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '8d /etc/   '}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_local_dir_pass - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_local_dir_pass - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ls', '-ltA', '--full-time', '/etc/']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_local_dir_pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'Activity_local_dir_pass', 'message': 'Activity_local_dir_pass  OK - local -    0.0 days  (   8 days  max)  /etc/'}

Test 2 - Local activity fail - CRITICAL --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_local_dir_fail - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_local_dir_fail', 'tag': 'local_dir_fail', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1s /etc/'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_local_dir_fail - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_local_dir_fail - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ls', '-ltA', '--full-time', '/etc/']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_local_dir_fail - eval_status() returned:  {'rslt': 3, 'notif_key': 'Activity_local_dir_fail', 'message': '  CRITICAL: Activity_local_dir_fail  STALE FILES - local - 1129.4 secs  (   1 secs  max)  /etc/'}

Test 3 - Remote directory activity pass - OK --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_remote_dir_pass - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_remote_dir_pass', 'tag': 'remote_dir_pass', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '5m /mnt/RAMDRIVE/   '}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_dir_pass - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_remote_dir_pass - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE/']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_dir_pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'Activity_remote_dir_pass', 'message': 'Activity_remote_dir_pass  OK - testhost2 -    0.0 mins  (   5 mins  max)  /mnt/RAMDRIVE/'}

Test 4 - Remote file activity pass - OK --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_remote_file_pass - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_remote_file_pass', 'tag': 'remote_file_pass', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '5m /mnt/RAMDRIVE/touchfile'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_file_pass - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_remote_file_pass - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_file_pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'Activity_remote_file_pass', 'message': 'Activity_remote_file_pass  OK - testhost2 -    0.0 mins  (   5 mins  max)  /mnt/RAMDRIVE/touchfile'}

Test 5 - Remote activity fail - CRITICAL --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_remote_dir_fail - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_remote_dir_fail', 'tag': 'remote_dir_fail', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1s /mnt/RAMDRIVE/'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_dir_fail - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_remote_dir_fail - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE/']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_remote_dir_fail - eval_status() returned:  {'rslt': 3, 'notif_key': 'Activity_remote_dir_fail', 'message': '  CRITICAL: Activity_remote_dir_fail  STALE FILES - testhost2 -    5.1 secs  (   1 secs  max)  /mnt/RAMDRIVE/'}

Test 6 - Local empty dir - CRITICAL --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_empty_dir - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_empty_dir', 'tag': 'empty_dir', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1h junkdir/'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_empty_dir - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_empty_dir - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ls', '-ltA', '--full-time', 'junkdir/']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_empty_dir - eval_status() returned:  {'rslt': 3, 'notif_key': 'Activity_empty_dir', 'message': '  CRITICAL: Activity_empty_dir  STALE FILES - local - junkdir/ no files found'}

Test 7 - Local no such file - CRITICAL --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_no_such_file - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_no_such_file', 'tag': 'no_such_file', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1h junkdir/xxx'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_no_such_file - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_no_such_file - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ls', '-ltA', '--full-time', 'junkdir']>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_no_such_file - eval_status() returned:  {'rslt': 3, 'notif_key': 'Activity_no_such_file', 'message': '  CRITICAL: Activity_no_such_file  STALE FILES - local - junkdir/xxx file not found'}

Test 8 - No such host - WARNING --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_Unknown_host - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_Unknown_host', 'tag': 'Unknown_host', 'host': 'nosuchhost', 'user_host_port': 'me@nosuchhost', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '1h /mnt/RAMDRIVE/xxx'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_Unknown_host - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_Unknown_host - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: Could not resolve hostname nosuchhost: Name or service not known> - attempting simplessh connection to <me@nosuchhost>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: Could not resolve hostname nosuchhost: Name or service not known>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_Unknown_host - eval_status() returned:  {'rslt': 1, 'notif_key': 'Activity_Unknown_host', 'message': '  WARNING: Activity_Unknown_host - nosuchhost - ssh: Could not resolve hostname nosuchhost: Name or service not known'}

Test 9 - Known host, unavailable - WARNING --------------------------------------------------
                  fsactivity_plugin.setup                -    DEBUG:  Activity_Unavailable_host - lanmonitor.fsactivity_plugin.setup() called:
  {'key': 'Activity_Unavailable_host', 'tag': 'Unavailable_host', 'host': 'testhostX', 'user_host_port': 'me@testhostX', 'critical': True, 'cmd_timeout': 4, 'check_interval': 1, 'rest_of_line': '1h /mnt/RAMDRIVE/xxx'}
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_Unavailable_host - setup() returned:  0
                  fsactivity_plugin.eval_status          -    DEBUG:  Activity_Unavailable_host - lanmonitor.fsactivity_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@testhostX', '-p22', '-T', 'ls', '-ltA', '--full-time', '/mnt/RAMDRIVE']>:  ssh: connect to host testhostx port 22: No route to host
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: connect to host testhostx port 22: No route to host> - attempting simplessh connection to <me@testhostX>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>:  ssh: connect to host testhostx port 22: No route to host
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: connect to host testhostx port 22: No route to host>
             fsactivity_plugin_test.dotest               -    DEBUG:  Activity_Unavailable_host - eval_status() returned:  {'rslt': 1, 'notif_key': 'Activity_Unavailable_host', 'message': '  WARNING: Activity_Unavailable_host - testhostX - ssh: connect to host testhostx port 22: No route to host'}
